### PR TITLE
[network] Add inbound rpc request protocol fuzzing

### PR DESCRIPTION
**Pull request by phlip9**
Summary:

+ Fuzzing attempts to target the `network::protocols::rpc`, `tokio::codec::Framed`, `unsigned_varint`, and `bytes::Bytes` codepaths.

+ The underlying tool, `cargo-fuzz`, relies on unstable compiler flags to
run fuzzers with sanitizers turned on. Adding `RUSTC_BOOTSTRAP=1` allows
unstable flags on stable compilers.

+ Adding `--release` flag enables optimizations, so fuzzers run ~3x
faster. Adding `--debug-assertions` ensures that we still check
`debug_assert!()` assertions.

+ Make `AccountAddress::new` a `const fn`

Test Plan:

```
$ cd testsuite/libra-fuzzer
$ RUSTC_BOOTSTRAP=1 cargo run fuzz inbound_rpc_protocol -- --release --debug-assertions
```

### CODE REVIEW

**Comment by bmwill** [comment]
It looks like this PR does a number of different, disjoint things, mind breaking up the single commit into 3 distinct ones?

File: network/Cargo.toml [hunk 1]
```diff
@@ -50,3 +57,8 @@ harness = false
 [[bench]]
 name = "network_bench"
 harness = false
+
+[features]
+default = []
+# fuzzing = ["proptest", "proptest_helpers", "tokio-codec"]
+fuzzing = ["proptest", "proptest_helpers"]
```
Comment (bmwill) [inline network/Cargo.toml]: :/ Hmm Have we finally made a call on what we're doing with these features? @huitseeker @metajack

File: network/Cargo.toml [hunk 2]
```diff
@@ -34,9 +34,16 @@ noise = { path = "noise" }
 socket_bench_server = { path = "socket_bench_server" }
 types = { path = "../types" }
 
+proptest = { version = "0.9.4", default-features = false, optional = true }
+proptest_helpers = { path = "../common/proptest_helpers", optional = true }
+# tokio-codec = { version = "0.1", optional = true }
```
Comment (bmwill) [inline network/Cargo.toml]: Do you intend for this to be commented out?

File: network/Cargo.toml [hunk 3]
```diff
@@ -34,9 +34,16 @@ noise = { path = "noise" }
...
+
 [dev-dependencies]
 criterion = "0.2.11"
 crypto = { path = "../crypto/crypto", features = ["testing"] }
+proptest = { version = "0.9.4", default-features = false }
+proptest_helpers = { path = "../common/proptest_helpers" }
+# tokio-codec = "0.1"
```
Comment (bmwill) [inline network/Cargo.toml]: same here

File: network/Cargo.toml [hunk 4]
```diff
@@ -50,3 +57,8 @@ harness = false
 [[bench]]
 name = "network_bench"
 harness = false
+
+[features]
+default = []
+# fuzzing = ["proptest", "proptest_helpers", "tokio-codec"]
```
Comment (bmwill) [inline network/Cargo.toml]: Did you intend for this to be commented out?

File: network/src/protocols/rpc/mod.rs [hunk 5]
```diff
@@ -89,6 +89,11 @@ pub mod utils;
 #[cfg(test)]
 mod test;
 
+#[cfg(any(feature = "fuzzing", test))]
+#[path = "fuzz.rs"]
+/// fuzzing module for the rpc protocol
+pub mod fuzz;
```
Comment (mimoo) [inline network/src/protocols/rpc/mod.rs]: we use this pattern with other fuzzers, maybe we should try and use the same name for the fuzzing module. No preference as to which naming pattern is better :P

File: testsuite/libra-fuzzer/README.md [hunk 6]
```diff
@@ -20,10 +20,14 @@ To be effective, fuzzing requires a corpus of existing inputs. This
 crate contains support for generating corpuses with `proptest`. Generate
 a corpus with `cargo run generate <target>`.
 
-Once a corpus has been generated, the fuzzer is ready to use: run
-`cargo run fuzz <target>`. 
+Once a corpus has been generated, the fuzzer is ready to use, simply run:
 
-For more options, run `cargo run -- --help`.
+```
+RUSTC_BOOTSTRAP=1 cargo run fuzz <target> -- --release --debug-assertions
```
Comment (mimoo) [inline testsuite/libra-fuzzer/README.md]: I can't find where I read that `debug-assertions` was disabling `--release`, in any case it shouldn't be bad as long that it's not `--release` that disables `--debug-assertions` :D

**Comment by mimoo** [comment]
if you address @bmwill 's comments it looks good to me! @bors-libra delegate+

File: network/src/protocols/rpc/mod.rs (hunk 5 above)
Comment (phlip9) [reply network/src/protocols/rpc/mod.rs]: do you mean like `event_processor_fuzzing.rs` vs `rpc/fuzz.rs`? would you prefer `rpc/fuzzing.rs` or `rpc/inbound_request_fuzzing.rs` or something else?

**Pull request closed without merging**